### ISSUE TITLE
sql: add memory accounting for MVCCScan requests on behalf of kv subsystem

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#51551]
One of the largest (if not *the* largest) under-accounted sources of memory allocation in CockroachDB is bytes allocated during `MVCCScan`, the primary way in which we retrieve data from storage into memory.

Under the covers, RocksDB or Pebble must allocate buffers for the data being read from disk, and 1-2 same-sized buffers must be allocated as the data percolates up the stack, depending on various factors that I haven't re-researched in a while. This memory is not accounted for by the system, and has been the root cause in several OOM issues lately. See #51548 for an example of potential pathological behavior caused by SQL's willingness to send `MVCCScan` queries with no regard for the underlying system's lack of backpressure for memory requests.

A proposed mitigation suggested by @ajwerner is to have SQL account for the memory use of the `MVCCScan` requests it issues, before issuing the requests. There are only a few significant places where SQL issues such requests, so it's a feasible ask. Now that KV exposes a bytes limit, SQL could request the full amount of that limit before sending the scan, which would protect the system against unbounded memory use.

A few downsides to this idea:

1. It's going to over-account memory in cases where the MVCCScan request returns just a small amount of data, like in the KV workload.
2. It's going to significantly change the meaning of `--max-sql-memory` in a way that might be confusing - the concept of "SQL Memory" will now include bytes returned from storage.
3. KV will not be able to control its own destiny. But, KV already does no accounting at all, so I think it's okay for SQL to do a bad approximation to protect the system from OOMs.

Looking for thoughts on this proposal.

cc @lunevalex @petermattis @knz @asubiotto @yuzefovich

**Comment by jordanlewis** [comment]
(also cc @joshimhoff @DuskEagle)

**Comment by joshimhoff** [comment]
WOOHOO

**Comment by knz** [comment]
no objection

**Comment by yuzefovich** [comment]
I'm in favor.

I think we could mitigate downside 1. by requesting the full bytes limit upfront and then adjusting the memory account according to the actual usage. The question are what is the life time of that memory account, how many accounts do we create - that will need to be thought through so that we don't have a big performance hit.

I'm less worried about downside 2. as long as we can document it well enough (for users who are interested in such details), and we'll also probably need to increase our recommended 25% value for such change.

And I agree with your position on downside 3 - unless there are any plans to introduce memory accounting below SQL, the stability should benefit from such change. If at some point we do add the memory accounting in KV or storage, then it should be a pretty simple removal from SQL layer.

**Comment by andreimatei** [comment]
> It's going to significantly change the meaning of --max-sql-memory in a way that might be confusing - the concept of "SQL Memory" will now include bytes returned from storage.

That's fine. "SQL Memory" is about memory used to serve queries, not about memory used by the sql code package or anything like that. In fact, what we actually want, I think, is a single `--memory-limit` flag without any qualifiers. And then separately limits for various particular caches, which would take out of the total memory limit.

**Comment by andreimatei** [comment]
I think that having SQL reserve memory is better than nothing, but a problem is that we'd possibly be reserving memory on the wrong node for a while. The scan responses live on leaseholders before making it to the gateway.
Now that Pebble is Go, it shouldn't be hard to actually have proper accounting throughout and let the storage engine reserve/account for what it actually reads. The scheme would be that the engine would increment some account for each scan, which account's ownership would then either be passed all the way to SQL if the local transport was used for the "RPC", or it'd be released when a `BatchResponse` has been passed to gRPC and another account would be opened by kvclient on the other side. Ideally we'd be able to hook into gRPC and peek at the size of the response and reserve memory before deserializing the proto. I assume this is possible somehow.

If @petermattis or friends bite and help on the storage side, I could try to help on the kv side.

**Comment by ajwerner** [comment]
>Now that Pebble is Go, it shouldn't be hard to actually have proper accounting throughout and let the storage engine reserve/account for what it actually reads. The scheme would be that the engine would increment some account for each scan, which account's ownership would then either be passed all the way to SQL if the local transport was used for the "RPC", or it'd be released when a BatchResponse has been passed to gRPC and another account would be opened by kvclient on the other side. Ideally we'd be able to hook into gRPC and peek at the size of the response and reserve memory before deserializing the proto. I assume this is possible somehow.

Do we really want to fail out requests? I'm not so concerned about the need to be precise in the accounting. The new scan-request byte limits should go a pretty long way. I'd rather see conservative, speculative resource accounting plus queuing. This all gets sort of complicated with regards to deadlock and queueing. Ideally we'd just have the KV layer hold on it its own memory monitor. We probably wouldn't want the KV layer's memory monitor to fail out requests but rather I think we'd want to queue. That runs into the various deadlock problems where we'll want to drop the acquired memory if we hit an intent. It doesn't seem that crazy and we've been talking about these problems in other contexts like rate-limiting tenants. There we accepted that when running in excess of a rate limit, we're okay potentially starving transaciton. I'm not yet sure we'd be willing to make the same claim for clusters which are getting squeezed on memory usage.

**Comment by petermattis** [comment]
> Now that Pebble is Go, it shouldn't be hard to actually have proper accounting throughout and let the storage engine reserve/account for what it actually reads.

I suspect that it isn't that Pebble is written in Go that enables this, but that `MVCCScan` is implemented in Go. Pebble itself uses very little memory (beyond `--cache`) for scan operations. The memory usage is in `MVCCScan` where we buffer up the results to return.

**Comment by andreimatei** [comment]
> Do we really want to fail out requests?

I think what I'd try to do is fail some requests (the same the SQL would fail according to this proposal) but not others. Like, anything that doesn't come from SQL, for example, or anything coming from system-class gRPC connection, I'd try to leave alone somehow.

**Comment by awoods187** [comment]
This seems like a big win for customers--provided we can figure out which knobs to provide and make it clear what they do. I'm partial to Andrei's point above on one limit that is comprised of further limits that are not visible to the user.

**Comment by jordanlewis** [comment]
I don't understand the discussion about failing queries being problematic. I think you *would* want to fail queries up front, no? We will have a really hard time building in queueing behavior on top of the current system, as well, and it will add a heap of complexity.

@andreimatei, your comment about the wrong node reserving memory doesn't seem particularly common to me, because while TableReaders *can* be planned on the wrong node, they usually aren't. Most of the time, the TableReader will be planned on the node that actually fetches the data. And, in the case that it is planned wrong, you *still* need to account for the memory that the TableReader will receive over the network. So in the worst case, you're always accounting for some real memory on the node that is making the request, but you're missing any accounting on the remote node that happens to serve the read. This is still a strict improvement over what we have today.

**Comment by andreimatei** [comment]
> because while TableReaders can be planned on the wrong node, they usually aren't. Most of the time, the TableReader will be planned on the node that actually fetches the data.

That's not true, is it? We don't distribute most queries. Back in the day we were only distributing full table scans. I hope we've gotten more sophisticated and distribute more of what looks like it might read a lot, but still there's plenty we don't distribute. And in multi-tenant cluster, we don't distribute any,

> And, in the case that it is planned wrong, you still need to account for the memory that the TableReader will receive over the network. So in the worst case, you're always accounting for some real memory on the node that is making the request, but you're missing any accounting on the remote node that happens to serve the read. This is still a strict improvement over what we have today.

Right, I think reserving memory would be a step forward. But it'd be that - "reserving" instead of "accounting" since it'd all be approximate and presumably we'd have to adjust the reservation once we actually got the data. I personally would try to actually build accounting from the MVCCScan up.

**Comment by tbg** [comment]
TL;DR: I think we should have something in TableReader and KV. TableReader asks for one byte when sending and reserves the rest when receiving. KV reserves accurately while it scans. In either case, failures to reserve fail the request. Both use the sql monitor local to the node.

----

Looks like the proposals discussed here are:

1. in SQL, account for `TargetBytes` before sending a scan and release it after getting the result, using the sql mem monitor.
2. in KV, account for `TargetBytes` when receiving a scan and release it "after" having responded. (i.e. same mechanism as 1, but pushed to the KV side). Could use the local sql mem monitor, or a new one.
3. in KV, account for data as it's read and release it "after" having responded. Again this could use the sql or a new monitor.
4. (the one I'm making towards the end of this post): 3 for KV and something "new" using courtesy byte in TableReader.

(combinations and smaller variations on those are possible, but this hopefully gives an idea).

To all, the discussion about what happens when the limit is exceeded applies: do we queue or fail? I personally empathize with the sentiment that adding queuing here is premature: it's complex, raises as many questions as it answers, and it's unclear that it solves the problem better in "relevant" ways as we don't expect a healthy cluster+workload to run into the limit anyway. Queuing also requires more work on the UX side as it needs to be explained and pointed out. This is on top of "walk before run" being a good idea here, anyway, so I will not consider queuing for now (but if someone else has strong feelings, please push back!).

As for the problem we're solving, based on Jordan's initial reference to https://github.com/cockroachdb/cockroach/issues/51548, it is my understanding that we're now defending against accidental scans with high concurrency. (Prior to the introduction of `TargetBytes`, the prime driver of OOMs were large scans, but those are now paginated). So this is either something silly as `UNION ALL` (the linked issue) or the client "accidentally" issuing 1000 full table scans in parallel (which with default `TargetBytes` would read the data in 10mb chunks each).

A large scan will occupy memory on both the KV and SQL nodes. If they are the same node, I *think* it would be occupied only once due to the local internal server shortcut, so we may have to do some elision here if we're being careful. (Also, of course the data is in general resident in memory at least twice, because of proto marshalling/unmarshalling, I will ignore that here).

Now on to the actual proposals. I'm diffusely anxious about 1+2 because we can be off by so much. As far as I can tell, the table readers can't generally tell what a reasonable `TargetBytes` will be. The default, 10mb, is slapped on anything with a limit:

https://github.com/cockroachdb/cockroach/blob/3dd344f309218b310a54673b6bbd56aa55c9374f/pkg/sql/row/kv_batch_fetcher.go#L244-L253

I think this could easily cause errors to pop up where there shouldn't be any, as we'd be reserving orders of magnitude too much. At the very least, the optimizer would have to provide estimates here, but even then, what if the table doesn't have good statistics or none at all? Note that the 10mb limit was chosen at introduction of `TargetBytes` to make sure we weren't regressing on any large scan performance. The choice was somewhat arbitrary, but essentially we can't just make this really small to "solve" the inaccuracy problem here.

At the same time, if we *don't* reserve anything here, we might get into trouble, as we will in fact be receiving data. Something that @ajwerner and I came up with in the context of multi-tenancy was the concept of a "courtesy byte": essentially, when issuing the request, we reserve one byte (certainly the request will use more, so this is undercounting). If that already fails, then we're at the limit and should not go ahead with the request. If it does succeed, we wait for the response, at which point we know the true size, and reserve `$true_size-1` additional bytes (nb: bit of a rat tail here - we  kind of want this to always succeed, as we already *have* the memory allocated). These are then released when the response goes out of scope (I guess when the rows are emitted, the fetcher releases the corresponding memory allocation as the row is now tracked higher up in SQL, not sure about the details). This isn't "hard" protection against OOMs, but in a garbage collected language there's no such thing anyway, and this seemed like a good approximation in the context in which we first used it. It seems likely that it is a decent fit here.

At the KV layer, I am sympathetic towards 3), assuming we have a single code path that does the aggregation. Right now, there is this possible delegation that can lead $anywhere:

https://github.com/cockroachdb/cockroach/blob/858541622a74121007f6759192e61f187eb7a8cd/pkg/storage/mvcc.go#L2347-L2350

I think (but @petermattis should confirm) that this branch is taken for RocksDB iterators only, which would imply that once RocksDB is out, we can remove this code and would be able to have central accounting logic in `pebbleMVCCScanner` which acquires chunks of memory from a mem monitor and fails when it runs out.
As for which memory monitor to use, I would tend to use the existing memory monitor. As a user, it's likely very hard to reason about which side of the boundary memory goes, and either way, it is reserved for "SQL query processing". I would rather explain to users that they may have to bump their `--sql-max-memory` flag than introduce another flag that is hard to understand and needs to be set to avoid accidentally oversubscribing.

**Comment by tbg** [comment]
cc @joshimhoff @jordanlewis I'd be interested in more examples where things went sideways, where TargetBytes alone didn't save the day.

**Comment by petermattis** [comment]
> I think (but @petermattis should confirm) that this branch is taken for RocksDB iterators only, which would imply that once RocksDB is out, we can remove this code and would be able to have central accounting logic in pebbleMVCCScanner which acquires chunks of memory from a mem monitor and fails when it runs out.

Yes, I believe that code branch is only used by RocksDB. `pebbleMVCCScanner` could allocate chunks of memory from a memory monitor. Do we need to fail when we hit the memory limit? I think we could return whatever we've scanned so far.

**Comment by jordanlewis** [comment]
While we're talking about this, @bdarnell also suggests adding accounting for the transaction objects whose sizes are governed by the `max_intent_bytes` setting.

**Comment by tbg** [comment]
> Do we need to fail when we hit the memory limit? I think we could return whatever we've scanned so far.

I would fail. Otherwise, we may end up accidentally implementing some inefficient form of queuing where instead of waiting, you bounce back and forth between the tablereader and KV (ask for 1000 rows, get 1 row, ask for 999 rows, get 0 rows, ask for 999 rows, get 1 row, ...); even if it's not quite as pathological, it will be hard to explain and expose this behavior properly. An error that reaches the client is much less likely to incur a lengthy debug session.

This is on top of returning *less* than requested being new behavior that could expose bugs. I think this concern could be ruled out, though, and `TargetBytes` largely already adds this behavior to scans, I think we're pretty good at just looking at the ResumeSpan.

**Comment by jordanlewis** [comment]
It's worth noting that the vectorized engine already does ahead-of-time memory reservations on behalf of KV. I'm going to add a similar reservation for the row engine for 20.2, just to protect us until a more sophisticated approach is added.
